File: db/alerts.py
"""
db.alerts – Warranty / licence expiry alert queries.
"""

from datetime import date, timedelta

from db.config import EXPIRY_WARNING_DAYS, TABLE_CONFIG
from db.connection import get_connection


def get_expiring_soon(days: int = EXPIRY_WARNING_DAYS) -> list[dict]:
    today = date.today().isoformat()
    threshold = (date.today() + timedelta(days=days)).isoformat()
    results: list[dict] = []
    with get_connection() as conn:
        for cat, cfg in TABLE_CONFIG.items():
            if cat == "Users":
                continue
            tbl = cfg["table"]
            exp_col = "expiry_date" if "expiry_date" in cfg["db_cols"] else "warranty_expiry"
            if exp_col not in cfg["db_cols"]:
                continue
            rows = conn.execute(
                f"SELECT *, '{cat}' AS type FROM {tbl} "
                f"WHERE {exp_col} BETWEEN ? AND ? AND status != 'Retired' "
                f"ORDER BY {exp_col}",
                (today, threshold),
            ).fetchall()
            results.extend(dict(r) for r in rows)
    return results


def get_expired() -> list[dict]:
    today = date.today().isoformat()
    results: list[dict] = []
    with get_connection() as conn:
        for cat, cfg in TABLE_CONFIG.items():
            if cat == "Users":
                continue
            tbl = cfg["table"]
            exp_col = "expiry_date" if "expiry_date" in cfg["db_cols"] else "warranty_expiry"
            if exp_col not in cfg["db_cols"]:
                continue
            rows = conn.execute(
                f"SELECT *, '{cat}' AS type FROM {tbl} "
                f"WHERE {exp_col} < ? AND status != 'Retired' "
                f"ORDER BY {exp_col}",
                (today,),
            ).fetchall()
            results.extend(dict(r) for r in rows)
    return results

File: internal/services/alerts.go
// Package services provides database access functions for the inventory application.
package services

import (
	"database/sql"
	"fmt"
	"strings"
	"time"

	"inventory/internal/models"
)

// alertSources lists every (table, date column) pair that contributes to
// expiry/warranty alerts. Mirrors db/alerts.py: Windows Key is intentionally
// omitted (no expiry column); Users is irrelevant.
type alertSource struct {
	table    string
	category string
	dateCol  string
}

var alertSources = []alertSource{
	{"computers", "Computer", "warranty_expiry"},
	{"smartphones", "Smartphone", "warranty_expiry"},
	{"tablets", "Tablet", "warranty_expiry"},
	{"antivirus", "Antivirus", "expiry_date"},
	{"other_software", "Other Software", "expiry_date"},
}

// GetAlerts returns expiry and warranty alerts for items expiring within warningDays.
func GetAlerts(db *sql.DB, warningDays int) ([]models.Alert, error) {
	threshold := fmt.Sprintf("+%d days", warningDays)
	parts := make([]string, 0, len(alertSources))
	params := make([]any, 0, len(alertSources))
	for _, src := range alertSources {
		parts = append(parts, fmt.Sprintf(`
		SELECT '%s' AS category, id, COALESCE(name, '') AS name, %s AS expiry_date,
		       CAST(julianday(%s) - julianday('now') AS INTEGER) AS days_remaining
		FROM %s
		WHERE %s IS NOT NULL
		  AND date(%s) <= date('now', ?)
		  AND status != 'Retired'`,
			src.category, src.dateCol, src.dateCol, src.table, src.dateCol, src.dateCol))
		params = append(params, threshold)
	}
	query := strings.Join(parts, "\n UNION ALL \n") + "\n ORDER BY days_remaining ASC"

	rows, err := db.Query(query, params...)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	today := time.Now().Format("2006-01-02")
	var out []models.Alert
	for rows.Next() {
		var a models.Alert
		if err := rows.Scan(&a.Category, &a.ID, &a.Name, &a.ExpiryDate, &a.DaysRemaining); err != nil {
			return nil, err
		}
		if a.ExpiryDate < today {
			a.Severity = "expired"
		} else {
			a.Severity = "expiring"
		}
		out = append(out, a)
	}
	return out, rows.Err()
}
